OmsAgent/omsagent.py

```diff
@@ -46,100 +49,279 @@
 if os.path.exists(ext_log_path):
     os.chmod(ext_log_path, 700)
 
+
 def main():
     waagent.LoggerInit('/var/log/waagent.log','/dev/stdout', True)
     waagent.Log("%s started to handle." %(ExtensionShortName))
 
+    # Determine the operation being executed
+    operation = None
     try:
         for a in sys.argv[1:]:
             if re.match("^([-/]*)(disable)", a):
-                hutil = parse_context("Disable")
-                disable(hutil)
+                operation = 'Disable'
             elif re.match("^([-/]*)(uninstall)", a):
-                hutil = parse_context("Uninstall")
-                uninstall(hutil)
+                operation = 'Uninstall'
             elif re.match("^([-/]*)(install)", a):
-                hutil = parse_context("Install")
-                install(hutil)
+                operation = 'Install'
             elif re.match("^([-/]*)(enable)", a):
-                hutil = parse_context("Enable")
-                enable(hutil)
+                operation = 'Enable'
             elif re.match("^([-/]*)(update)", a):
-                dummy_command("Update", "success", "Update succeeded")
+                operation = 'Update'
     except Exception as e:
-        err_msg = ("Failed to enable the extension with error: {0}, "
-                   "{1}").format(e, traceback.format_exc())
-        waagent.Error(err_msg)
-        hutil.error(err_msg)
-        hutil.do_exit(1, 'Enable','failed','0',
-                      'Enable failed: {0}'.format(e))
+        waagent.Error(e.message)
+
+    # Set up for exit code and any error messages
+    exit_code = 0
+    message = operation + ' succeeded'
+
+    # Invoke operation
+    try:
+        hutil = parse_context(operation)
+        if is_vm_supported_for_extension():
+            exit_code = operations[operation](hutil)
+        else:
+            log_and_exit(hutil, operation, 51, 'Unsupported operation system')
+
+        if exit_code is not 0:
+            message = (operation + ' failed with exit code {0}').format(exit_code)
+
+    except ValueError as e:
```

Comment:
OK, I saw you are raising ValueError in enable(), so that's probably why you are catching only ValueError here, but the parse_context() may throw KeyError if necessary JSON key is not present in either the privateSettings or the publicSettings, so I'd do the same on KeyError as well.